### PR TITLE
Allow extensions to contribute to WorkspaceService

### PR DESCRIPTION
**Pull request by mickaelistria**
Currently supports
didChangeWorkspaceFolders

Signed-off-by: Mickael Istria <mistria@redhat.com>

### CODE REVIEW

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/IWorkspaceServiceParticipant.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ *  which accompanies this distribution, and is available at
+ *  http://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.lemminx.services.extensions;
+
+import org.eclipse.lsp4j.DidChangeWorkspaceFoldersParams;
+
+public interface IWorkspaceServiceParticipant {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/IWorkspaceServiceParticipant.java:14]: Add Javadoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/IWorkspaceServiceParticipant.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+	public void didChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams params);
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/IWorkspaceServiceParticipant.java:16]: Add Javadoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/XMLExtensionsRegistry.java [hunk 3]
```diff
@@ -32,6 +32,7 @@
 import org.eclipse.lemminx.services.extensions.save.ISaveContext.SaveContextType;
 import org.eclipse.lemminx.uriresolver.URIResolverExtensionManager;
 import org.eclipse.lsp4j.InitializeParams;
+import org.eclipse.lsp4j.services.WorkspaceService;
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/XMLExtensionsRegistry.java:35]: should be not removed?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/XMLExtensionsRegistry.java [hunk 4]
```diff
@@ -196,6 +199,15 @@ public void doSave(ISaveContext saveContext) {
 		return symbolsProviderParticipants;
 	}
 
+	/**
+	 * @return
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/XMLExtensionsRegistry.java:203]: Add Javadoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/XMLExtensionsRegistry.java [hunk 5]
```diff
@@ -356,6 +368,22 @@ public void registerSymbolsProviderParticipant(ISymbolsProviderParticipant symbo
 	public void unregisterSymbolsProviderParticipant(ISymbolsProviderParticipant symbolsProviderParticipant) {
 		symbolsProviderParticipants.remove(symbolsProviderParticipant);
 	}
+	
+	/**
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/XMLExtensionsRegistry.java:372]: Add Javadoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/XMLExtensionsRegistry.java [hunk 6]
```diff
@@ -356,6 +368,22 @@ public void registerSymbolsProviderParticipant(ISymbolsProviderParticipant symbo
...
+	 * @param workspaceServiceParticipant
+	 * @since 0.14.2
+	 */
+	public void registerWorkspaceServiceParticipant(IWorkspaceServiceParticipant workspaceServiceParticipant) {
+		workspaceServiceParticipants.add(workspaceServiceParticipant);
+	}
+
+	/**
+	 * @param workspaceServiceParticipant
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/XMLExtensionsRegistry.java:381]: Add Javadoc

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/extensions/WorkspaceServiceParticipantTest.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+			this.didChangeWorkspaceFolders = params;
+		}
+	}
+	
+	private CaptureWokspaceServiceCalls workspaceServiceParticipant;
+	private XMLLanguageServer server;
+
+	@BeforeEach
+	public void initializeLanguageService() {
+		this.server = new XMLLanguageServer();
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/extensions/WorkspaceServiceParticipantTest.java:46]: write just:

`this.server = new MockXMLLanguageServer()`

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/extensions/WorkspaceServiceParticipantTest.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import org.eclipse.lsp4j.MessageParams;
+import org.eclipse.lsp4j.PublishDiagnosticsParams;
+import org.eclipse.lsp4j.RegistrationParams;
+import org.eclipse.lsp4j.ShowMessageRequestParams;
+import org.eclipse.lsp4j.WorkspaceFolder;
+import org.eclipse.lsp4j.WorkspaceFoldersChangeEvent;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class WorkspaceServiceParticipantTest {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/extensions/WorkspaceServiceParticipantTest.java:29]: Add Javadoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/IWorkspaceServiceParticipant.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
+/**
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/IWorkspaceServiceParticipant.java:1]: Please add contributors in the header:

```
/*******************************************************************************
* Copyright (c) 2021 Red Hat Inc. and others.
* All rights reserved. This program and the accompanying materials
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Contributors:
*     Red Hat Inc. - initial API and implementation
*******************************************************************************/
```

Comment (mickaelistria) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/IWorkspaceServiceParticipant.java:1]: FWIW, it's been clarified by EMO that according to latest version of EDP those Contributors block are not at all mandatory for Eclipse projects.

**Comment by mickaelistria** [comment]
I think my last patch does cover all concerns you expressed.

**Comment by angelozerr** [comment]
Thanks @mickaelistria !

**Comment by mickaelistria** [comment]
Thanks a lot for the review! Do you have an estimation of when is next release going to happen? Depending on the release date, we may merge the LemMinX-Maven change leveraging this API right now or later.

**Comment by angelozerr** [comment]
> Thanks a lot for the review! 

You are welcome!

> Do you have an estimation of when is next release going to happen? Depending on the release date, we may merge the LemMinX-Maven change leveraging this API right now or later.

We will do a release as soon as the binary stuff will be available by @datho7561

**Pull request merged**